cudax/include/cuda/experimental/__memory_resource/cuda_memory_pool.h

```diff
@@ -0,0 +1,270 @@
+    __pool_properties.location.type = ::cudaMemLocationTypeDevice;
+    __pool_properties.location.id   = __device_id;
+    ::cudaMemPool_t __cuda_pool_handle{};
+    _CCCL_TRY_CUDA_API(::cudaMemPoolCreate, "Failed to call cudaMemPoolCreate", &__cuda_pool_handle, &__pool_properties);
+
+    // CUDA drivers before 11.5 have known incompatibilities with the async allocator.
+    // We'll disable `cudaMemPoolReuseAllowOpportunistic` if cuda driver < 11.5.
+    // See https://github.com/NVIDIA/spark-rapids/issues/4710.
+    int __driver_version = 0;
+    _CCCL_TRY_CUDA_API(::cudaDriverGetVersion, "Failed to call cudaDriverGetVersion", &__driver_version);
+
+    constexpr int __min_async_version = 11050;
+    if (__driver_version < __min_async_version)
+    {
+      int __disable_reuse = 0;
+      _CCCL_TRY_CUDA_API(
+        ::cudaMemPoolSetAttribute,
+        "Failed to call cudaMemPoolSetAttribute with cudaMemPoolReuseAllowOpportunistic",
+        __cuda_pool_handle,
+        ::cudaMemPoolReuseAllowOpportunistic,
+        &__disable_reuse);
+    }
+
+    _CCCL_TRY_CUDA_API(
+      ::cudaMemPoolSetAttribute,
+      "Failed to call cudaMemPoolSetAttribute with cudaMemPoolAttrReleaseThreshold",
+      __cuda_pool_handle,
+      ::cudaMemPoolAttrReleaseThreshold,
+      &__properties.release_threshold);
+
+    // allocate the requested initial size to prime the pool.
+    // We need to use a new stream so we do not wait on other work
+    if (__properties.initial_pool_size != 0)
+    {
+      ::cuda::experimental::stream __temp_stream{__device_id};
+      void* __ptr{nullptr};
+      _CCCL_TRY_CUDA_API(
+        ::cudaMallocAsync,
+        "cuda_memory_pool failed to allocate the initial pool size",
+        &__ptr,
+        __properties.initial_pool_size,
+        __temp_stream.get());
+
+      _CCCL_ASSERT_CUDA_API(
+        ::cudaFreeAsync, "cuda_memory_pool failed to free the initial pool allocation", __ptr, __temp_stream.get());
+    }
+    return __cuda_pool_handle;
+  }
+
+public:
+  //! @brief Constructs a \c cuda_memory_pool with the optionally specified initial pool size and release threshold.
+  //! If the pool size grows beyond the release threshold, unused memory held by the pool will be released at the next
+  //! synchronization event.
+  //! @throws cuda_error if the CUDA version does not support ``cudaMallocAsync``.
+  //! @param __device_id The device id of the device the stream pool is constructed on.
+  //! @param __pool_properties Optional, additional properties of the pool to be created.
+  explicit cuda_memory_pool(const ::cuda::experimental::device_ref __device_id,
+                            cuda_memory_pool_properties __properties = {})
+      : __pool_handle_(__create_cuda_mempool(__device_id.get(), __properties))
+  {}
+
+  //! @brief Disables construction from a plain `cudaMemPool_t`. We want to ensure clean ownership semantics.
+  cuda_memory_pool(::cudaMemPool_t) = delete;
```

Comment:
Should there be `from_native_handle` static member function that transfers the ownership, like for stream, event?

Reply:
Makes sense, added